### PR TITLE
persist: finish porting over microbenchmarks

### PR DESCRIPTION
**Pull request by danhhz**
See individual commits for details.

This also introduces the MZ_PERSIST_BENCH_THROUGHPUT env for a quick,
rough estimate of throughput. Real measurement of throughput should
continue to be done with the open-loop benchmark.

### Motivation

   * This PR refactors existing code.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

File: src/persist-client/benches/benches.rs
Comment (danhhz) [inline src/persist-client/benches/benches.rs:62]: One of my motivations for the new structure was that the old persist benchmarks had started to become a bit of a jumble. Concretely, there were a number of them that only existed to help tune a specific PR and then stuck around, adding cruft everything we needed to benchmarks to evaluate a change. Here, I'm trying to break them up first roughly into 2 categories:
- "speed_of_light" (accepting name suggestions) which is a measure of fundamental things that persist's latency and throughput are dependent on. This basically means BlobMulti and Consensus impls, though I also tossed batch encoding in here. The latter is a bit different (it's indeed something we control) but it's also a sort of fundamentally separate thing from the rest of persist, especially with aljoscha's PR to spearate out a BatchBuilder in our interface.
- Then a group of things that are roughly "what our users see". They're called "min_latency" because that's what they were called before this PR and I haven't come up with any names that I like for this (again accepting suggestions).

I'm hoping that going forward, we can start establishing easily targeted (on the cargo bench cli) subsets of these for common situations. Walking through a couple examples:
- A PR _might_ affect latencies: `min_latency/*/mem_mem` will tell whether this makes a measurable impact at all. If it does, then as a followup `min_latency/*/s3_pg` tells whether it shows up in prod and `min_latency/*/file_pg` ditto for dev.
- We want to get a sense for persist's current prod latencies (say to update the README): `min_latency/*/s3_pg`
- We want to figure out how much overhead persist gives over the fundamental pieces we have no control over: `min_latency/writes/s3_pg` vs `speed_of_light/blob_set/s3` + `speed_of_light/consensus_compare_and_set/pg`
- We want a sense for what our max throughput is for open-loop: MZ_PERSIST_BENCH_THROUGHPUT=1 + `speed_of_light/blob_set/s3`
- Evaluating perf work within an individual BlobMulti or Consensus impl: `speed_of_light/*/<foo>`

One takeaway I'm getting from writing this comment up is that I dunno if `cargo test` supports the `*` wildcard, so we probably want to move the mem, file, s3, sqlite, pg stuff to be the second slot, right after `min_latency` or `speed_of_light`. I'll hold off until we get some PR comments here first, so I can make a batch of changes at once

**Comment by ruchirK** [comment]
reviewing now!

**Review by ruchirK — APPROVED** [review]
LGTM! I really like the philosophy here and think that having well defined subsets of microbenchmarks will help a lot. I don't have good naming suggestions but I do kind of think `min_latency` and `speed_of_light` seem more like synonyms than two distinct categories.

I skimmed the code changes and didn't review them too much in depth because the code changes seemed fairly straightforward. Let me know if there's something you want me to revisit in greater detail.

**Comment by aljoscha** [comment]
How do these two results fit together:

```
        min_latency/snapshot/file_pg/50MiB
                                time:   [256.18 ms 258.11 ms 260.19 ms]
                                thrpt:  [192.16 MiB/s 193.72 MiB/s 195.18 MiB/s]
        speed_of_light/blob_set/file/50MiB
                                time:   [399.96 ms 400.09 ms 400.22 ms]
                                thrpt:  [124.93 MiB/s 124.97 MiB/s 125.01 MiB/s]
```

I might be dense, but doesn't this say that snapshotting (which includes writing a blob) is faster than writing a blob? The other backends (mem, s3) don't have this anomaly.

**Review by aljoscha — APPROVED** [review]

**Comment by aljoscha** [comment]
I can offer `planck_constant` instead of `speed_of_light`. 😅 

Things that might work for realz:
 - raw vs end_to_end (though end-to-end is also a bit loaded already)
 - plumbing vs porcelain (hello git!)

**Comment by aljoscha** [comment]
> The snapshot bench is reading a snapshot, not writing one :)

Yeah, I noticed, but maybe you didn't see my edit. :sweat_smile:

**Comment by danhhz** [comment]
hmm immediately after I sent that, I realized that plumbing/porcelain might be growing on me. one could argue that the location stuff (which limited by bandwidth, which is what I've been focusing on) also is code we own, just at a lower (plumbing) level. that is actually more accurate for encode_batch than my "speed of light" intuition

**Comment by ruchirK** [comment]
plumbing vs porcelain is fine for me but can someone explain the metaphor in a few sentences so I'm sure I'm on the same page? sorry

**Comment by danhhz** [comment]
well ./bin/scratch suddenly isn't working for me, so I can't look into the discrepancy that aljoscha pointed out. I'll go ahead and merge this to unblock my other PR, but this is definitely something we'll look into at some point.

TFTRs!!

**Comment by danhhz** [comment]
@aljoscha I managed to get bin/scratch working again and this repro'd (!?!?)

```
porcelain/writes/file_pg/50MiB
                        time:   [67.246 ms 67.473 ms 67.712 ms]
                        thrpt:  [738.42 MiB/s 741.04 MiB/s 743.54 MiB/s]
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe
plumbing/blob_set/file/50MiB
                        time:   [399.98 ms 400.13 ms 400.29 ms]
                        thrpt:  [124.91 MiB/s 124.96 MiB/s 125.01 MiB/s]
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low severe
  1 (1.00%) low mild
  5 (5.00%) high severe
```

**Comment by aljoscha** [comment]
😱

**Pull request merged**
